File: src/main/java/analyzer/AnalyzerUtils.java
package analyzer;

import entity.InterfaceModel;
import entity.PacketModel;
import entity.PortModel;
import org.pcap4j.core.PcapNativeException;
import org.pcap4j.core.PcapNetworkInterface;
import org.pcap4j.core.Pcaps;
import org.pcap4j.packet.*;
import org.pcap4j.packet.namednumber.EtherType;
import org.pcap4j.packet.namednumber.IpNumber;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class AnalyzerUtils {
    public static List<InterfaceModel> getAllNetworkInterfacesName() throws PcapNativeException {
        List<PcapNetworkInterface> allInterFace = Pcaps.findAllDevs();
        List<InterfaceModel> interfaceModels = new ArrayList<>();
        int id = 0;
        for (PcapNetworkInterface networkInterface : allInterFace) {
            interfaceModels.add(new InterfaceModel(id++, networkInterface.getName()));
        }
        return interfaceModels;
    }

    public static PacketModel convert(Packet packet, long id) {
        if (packet.contains(EthernetPacket.class)) {
            EthernetPacket ethernetPacket = packet.get(EthernetPacket.class);
            long date = getTime();
            int size = packet.length();
            String descriptor = packet.toString();
            String srcMac = ethernetPacket.getHeader().getSrcAddr().toString();
            String dstMac = ethernetPacket.getHeader().getDstAddr().toString();

            String protocol;
            String srcIp;
            String dstIp;

            PortModel srcPort;
            PortModel dstPort;
            String extInfo;

            if (EtherType.IPV4.equals(ethernetPacket.getHeader().getType())) {
                IpV4Packet ipv4 = ethernetPacket.getPayload().get(IpV4Packet.class);
                protocol = ipv4.getHeader().getProtocol().name();
                srcIp = ipv4.getHeader().getSrcAddr().getHostAddress();
                dstIp = ipv4.getHeader().getDstAddr().getHostAddress();
                if (IpNumber.TCP.equals(ipv4.getHeader().getProtocol())) {
                    TcpPacket tcp = ipv4.getPayload().get(TcpPacket.class);
                    srcPort = new PortModel(tcp.getHeader().getSrcPort().valueAsInt(), tcp.getHeader().getSrcPort().name());
                    dstPort = new PortModel(tcp.getHeader().getDstPort().valueAsInt(), tcp.getHeader().getDstPort().name());
                    extInfo = tcp.getPayload().toString();
                } else if (IpNumber.UDP.equals(ipv4.getHeader().getProtocol())) {
                    UdpPacket udp = ipv4.getPayload().get(UdpPacket.class);
                    srcPort = new PortModel(udp.getHeader().getSrcPort().valueAsInt(), udp.getHeader().getSrcPort().name());
                    dstPort = new PortModel(udp.getHeader().getDstPort().valueAsInt(), udp.getHeader().getDstPort().name());
                    extInfo = udp.getPayload().toString();
                } else if (IpNumber.ICMPV4.equals(ipv4.getHeader().getProtocol())) {
                    IcmpV4CommonPacket icmp = ipv4.getPayload().get(IcmpV4CommonPacket.class);
                    srcPort = null;
                    dstPort = null;
                    extInfo = "type: " + icmp.getHeader().getType().toString() + " / code: " + icmp.getHeader().getCode().toString() + "\n" + icmp.getPayload().toString();
                } else if (IpNumber.IGMP.equals(ipv4.getHeader().getProtocol())) {
                    //TODO: Find new way to handle IGMP packets
                    srcPort = null;
                    dstPort = null;
                    extInfo = ipv4.getPayload().toString();
                } else {
                    srcPort = null;
                    dstPort = null;
                    extInfo = ipv4.getPayload().toString();
                }
            } else if (EtherType.IPV6.equals(ethernetPacket.getHeader().getType())) {
                IpV6Packet ipv6 = ethernetPacket.getPayload().get(IpV6Packet.class);
                protocol = ipv6.getHeader().getProtocol().name();
                srcIp = ipv6.getHeader().getSrcAddr().getHostAddress();
                dstIp = ipv6.getHeader().getDstAddr().getHostAddress();
                if (IpNumber.TCP.equals(ipv6.getHeader().getProtocol())) {
                    TcpPacket tcp = ipv6.getPayload().get(TcpPacket.class);
                    srcPort = new PortModel(tcp.getHeader().getSrcPort().valueAsInt(), tcp.getHeader().getSrcPort().name());
                    dstPort = new PortModel(tcp.getHeader().getDstPort().valueAsInt(), tcp.getHeader().getDstPort().name());
                    extInfo = tcp.getPayload().toString();
                } else if (IpNumber.UDP.equals(ipv6.getHeader().getProtocol())) {
                    UdpPacket udp = ipv6.getPayload().get(UdpPacket.class);
                    srcPort = new PortModel(udp.getHeader().getSrcPort().valueAsInt(), udp.getHeader().getSrcPort().name());
                    dstPort = new PortModel(udp.getHeader().getDstPort().valueAsInt(), udp.getHeader().getDstPort().name());
                    extInfo = udp.getPayload().toString();
                } else if (IpNumber.ICMPV6.equals(ipv6.getHeader().getProtocol())) {
                    IcmpV6CommonPacket icmp = ipv6.getPayload().get(IcmpV6CommonPacket.class);
                    srcPort = null;
                    dstPort = null;
                    extInfo = "type: " + icmp.getHeader().getType().toString() + " / code: " + icmp.getHeader().getCode().toString() + "\n" + icmp.getPayload().toString();
                } else {
                    srcPort = null;
                    dstPort = null;
                    extInfo = ipv6.getPayload().toString();
                }
            } else if (EtherType.ARP.equals(ethernetPacket.getHeader().getType())) {
                protocol = EtherType.ARP.name();
                ArpPacket arp = ethernetPacket.getPayload().get(ArpPacket.class);
                srcIp = arp.getHeader().getSrcProtocolAddr().getHostAddress();
                srcMac = arp.getHeader().getSrcHardwareAddr().toString();
                dstIp = arp.getHeader().getDstProtocolAddr().getHostAddress();
                dstMac = arp.getHeader().getDstHardwareAddr().toString();
                srcPort = null;
                dstPort = null;
                extInfo = "operation: " + arp.getHeader().getOperation().name() + "\n" + arp.getPayload();
//            } else if (EtherType.RARP.equals(ethernetPacket.getHeader().getType())) {
//            } else if (EtherType.DOT1Q_VLAN_TAGGED_FRAMES.equals(ethernetPacket.getHeader().getType())) {
//            } else if (EtherType.APPLETALK.equals(ethernetPacket.getHeader().getType())) {
//            } else if (EtherType.PPP.equals(ethernetPacket.getHeader().getType())) {
//            } else if (EtherType.MPLS.equals(ethernetPacket.getHeader().getType())) {
//            } else if (EtherType.PPPOE_DISCOVERY_STAGE.equals(ethernetPacket.getHeader().getType())) {
//            } else if (EtherType.PPPOE_SESSION_STAGE.equals(ethernetPacket.getHeader().getType())) {
            } else {
                protocol = ethernetPacket.getHeader().getType().name();
                srcIp = "";
                dstIp = "";
                srcPort = null;
                dstPort = null;
                extInfo = ethernetPacket.toString();
            }
            return new PacketModel(
                    id,
                    date,
                    size,
                    protocol,
                    srcMac,
                    srcIp,
                    srcPort,
                    dstMac,
                    dstIp,
                    dstPort,
                    extInfo,
                    descriptor
            );
        }
        return null;
    }

    private static long getTime() {
        return new Date().getTime();
    }
}

File: src/main/java/analyzer/AnalyzerUtils.kt
package analyzer

import entity.PortModel
import entity.PacketModel
import org.pcap4j.core.Pcaps
import org.pcap4j.packet.*
import org.pcap4j.packet.namednumber.EtherType
import org.pcap4j.packet.namednumber.IpNumber
import java.util.*

class AnalyzerUtils {
    companion object {

        fun getAllNetworkInterfacesName(): List<String> {
            return Pcaps.findAllDevs().map { it.name }
        }

        fun Packet.convert(id: Long): PacketModel? {
            if (contains(EthernetPacket::class.java)) {
                val ethernetPacket = get(EthernetPacket::class.java)
                val date = getTime()
                val size = length()
                val descriptor = toString()
                val srcMac = ethernetPacket.header.srcAddr.toString()
                val dstMac = ethernetPacket.header.dstAddr.toString()

                val protocol: String
                val srcIp: String
                val dstIp: String

                val srcPort: PortModel?
                val dstPort: PortModel?
                val extInfo: String

                when (ethernetPacket.header.type) {
                    EtherType.IPV4 -> {
                        val ipv4 = ethernetPacket.get(IpV4Packet::class.java).header
                        protocol = ipv4.protocol.name()
                        srcIp = ipv4.srcAddr.hostAddress
                        dstIp = ipv4.dstAddr.hostAddress
                        when (ipv4.protocol) {
                            IpNumber.TCP -> {
                                val tcp = ethernetPacket.get(TcpPacket::class.java)
                                tcp.header.also {
                                    srcPort = PortModel(it.srcPort.valueAsInt(), it.srcPort.name())
                                    dstPort = PortModel(it.dstPort.valueAsInt(), it.dstPort.name())
                                }
                                extInfo = tcp.payload.toString()
                            }
                            IpNumber.UDP -> {
                                val udp = ethernetPacket.get(UdpPacket::class.java)
                                udp.header.also {
                                    srcPort = PortModel(it.srcPort.valueAsInt(), it.srcPort.name())
                                    dstPort = PortModel(it.dstPort.valueAsInt(), it.dstPort.name())
                                }
                                extInfo = udp.payload.toString()
                            }
                            IpNumber.ICMPV4 -> {
                                val icmp = ethernetPacket.get(IcmpV4CommonPacket::class.java)
                                srcPort = null
                                dstPort = null
                                extInfo = "type: ${icmp.header.type} / code: ${icmp.header.code}\n${icmp.payload}"
                            }
                            IpNumber.IGMP -> {
                                srcPort = null
                                dstPort = null
                                extInfo = ethernetPacket.payload.toString()
                            }
                            else -> {
                                srcPort = null
                                dstPort = null
                                extInfo = ethernetPacket.payload.toString()
                            }
                        }
                    }
                    EtherType.IPV6 -> {
                        val ipv6 = ethernetPacket.get(IpV6Packet::class.java).header
                        protocol = ipv6.protocol.name()
                        srcIp = ipv6.srcAddr.hostAddress
                        dstIp = ipv6.dstAddr.hostAddress
                        when (ipv6.protocol) {
                            IpNumber.TCP -> {
                                val tcp = ethernetPacket.get(TcpPacket::class.java)
                                tcp.header.also {
                                    srcPort = PortModel(it.srcPort.valueAsInt(), it.srcPort.name())
                                    dstPort = PortModel(it.dstPort.valueAsInt(), it.dstPort.name())
                                }
                                extInfo = tcp.payload.toString()
                            }
                            IpNumber.UDP -> {
                                val udp = ethernetPacket.get(UdpPacket::class.java)
                                udp.header.also {
                                    srcPort = PortModel(it.srcPort.valueAsInt(), it.srcPort.name())
                                    dstPort = PortModel(it.dstPort.valueAsInt(), it.dstPort.name())
                                }
                                extInfo = udp.payload.toString()
                            }
                            IpNumber.ICMPV6 -> {
                                val icmp = ethernetPacket.get(IcmpV6CommonPacket::class.java)
                                srcPort = null
                                dstPort = null
                                extInfo = "type: ${icmp.header.type} / code: ${icmp.header.code}\n${icmp.payload}"
                            }
                            else -> {
                                srcPort = null
                                dstPort = null
                                extInfo = ethernetPacket.payload.toString()
                            }
                        }
                    }
                    EtherType.ARP -> {
                        protocol = EtherType.ARP.name()
                        val arp = ethernetPacket.get(ArpPacket::class.java)
                        srcIp = arp.header.srcProtocolAddr.hostAddress
                        dstIp = arp.header.dstProtocolAddr.hostAddress
                        srcPort = null
                        dstPort = null
                        extInfo = "operation: ${arp.header.operation.name()}\n${arp.payload}"
                    }
//                    EtherType.RARP -> {
//                    }
//                    EtherType.DOT1Q_VLAN_TAGGED_FRAMES -> {
//                    }
//                    EtherType.APPLETALK -> {
//                    }
//                    EtherType.PPP -> {
//                    }
//                    EtherType.MPLS -> {
//                    }
//                    EtherType.PPPOE_DISCOVERY_STAGE -> {
//                    }
//                    EtherType.PPPOE_SESSION_STAGE -> {
//                    }
                    else -> {
                        protocol = ethernetPacket.header.type.name()
                        srcIp = ""
                        dstIp = ""
                        srcPort = null
                        dstPort = null
                        extInfo = ethernetPacket.toString()
                    }
                }
                return PacketModel(
                        id,
                        date,
                        size,
                        protocol,
                        srcMac,
                        srcIp,
                        srcPort,
                        dstMac,
                        dstIp,
                        dstPort,
                        extInfo,
                        descriptor
                )
            }
            return null
        }

        private fun getTime(): Long {
            return Date().time
        }
    }
}